python/strings_udf/cpp/include/cudf/strings/udf/udf_string.cuh

```diff
@@ -0,0 +1,475 @@
+}
+
+__device__ inline udf_string& udf_string::append(cudf::char_utf8 chr, cudf::size_type count)
+{
+  if (count <= 0) { return *this; }
+  auto const char_bytes = cudf::strings::detail::bytes_in_char_utf8(chr) * count;
+  auto const nbytes     = m_bytes + char_bytes;
+  if (nbytes > m_capacity) { reallocate(2 * nbytes); }
+  auto out_ptr = m_data + m_bytes;
+  for (auto idx = 0; idx < count; ++idx) {
+    out_ptr += cudf::strings::detail::from_char_utf8(chr, out_ptr);
+  }
+  m_bytes         = nbytes;
+  m_data[m_bytes] = 0;
+  return *this;
+}
+
+__device__ inline udf_string& udf_string::append(cudf::string_view const in)
+{
+  return append(in.data(), in.size_bytes());
+}
+
+__device__ inline udf_string& udf_string::operator+=(cudf::string_view const in)
+{
+  return append(in);
+}
+
+__device__ inline udf_string& udf_string::operator+=(cudf::char_utf8 chr) { return append(chr); }
+
+__device__ inline udf_string& udf_string::operator+=(char const* str) { return append(str); }
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos,
+                                                 char const* str,
+                                                 cudf::size_type in_bytes)
+{
+  return replace(pos, 0, str, in_bytes);
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos, char const* str)
+{
+  return insert(pos, str, detail::bytes_in_null_terminated_string(str));
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos, cudf::string_view const in)
+{
+  return insert(pos, in.data(), in.size_bytes());
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos,
+                                                 cudf::size_type count,
+                                                 cudf::char_utf8 chr)
+{
+  return replace(pos, 0, count, chr);
+}
+
+__device__ inline udf_string udf_string::substr(cudf::size_type pos, cudf::size_type count) const
+{
+  if (pos < 0) { return udf_string{"", 0}; }
+  auto const spos = byte_offset(pos);
+  if (spos >= m_bytes) { return udf_string{"", 0}; }
+  auto const epos = count < 0 ? m_bytes : std::min(byte_offset(pos + count), m_bytes);
+  return udf_string{data() + spos, epos - spos};
+}
+
+// utility for replace()
+__device__ void udf_string::shift_bytes(cudf::size_type spos,
+                                        cudf::size_type epos,
+                                        cudf::size_type nbytes)
+{
+  if (nbytes < m_bytes) {
+    // shift bytes to the left [...wxyz] -> [wxyzxyz]
+    auto src = epos;
+    auto tgt = spos;
+    while (tgt < nbytes) {
+      m_data[tgt++] = m_data[src++];
+    }
+  } else if (nbytes > m_bytes) {
+    // shift bytes to the right [abcd...] -> [abcabcd]
+    auto src = m_bytes;
+    auto tgt = nbytes;
+    while (src > epos) {
+      m_data[--tgt] = m_data[--src];
+    }
+  }
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
```

Comment:
We document that this function does nothing if `pos > length()`, but we don't document what happens if `pos < length() && (pos + in_bytes) > length()`. We should add that.

Reply:
Sorry, I don't understand this request.

Reply:
I used the wrong variable in my example, which is probably adding to the confusion. I think I had two questions, of which only one is still relevant.

First question: what happens if overwriting`pos + count` characters leads to out of bounds accesses? Plugging the snippet below into [the compiler explorer](https://godbolt.org/z/YznPcdn8e) suggests that the code prevents out of bounds accesses. Is that something we should promise for our implementation too? 
```
#include <iostream>
#include <string>
  
int main()
{
    std::string str{"First"};
    // This line has the same behavior for any value of count >= 5. Is that expected 
    auto count = 100;
    str.replace(0, count, "Second");
    std::cout << str << std::endl;
}
``` 

Second question: what happens if the replacement string is larger than `count`? That use case seems like it is fine and the reallocation is the behavior expected by the standard. I don't know if we need to document that since on second thought it seems fairly obvious that this would be standard behavior since otherwise the API would be very limited if it could only replace a string of equal length.

Reply:
I think this is a reasonable expectation requiring no special documentation.